### ISSUE TITLE
Clarify documentation of the model property

### ISSUE THREAD

**Issue by thomaseizinger** [uber-hypermedia/specification#34]
Currently, the specification states the following for the model property:

`Contains an [RFC6570]-compliant string to be used to construct message bodies. Variables in UBER model strings SHOULD be resolved using the values from name properties, but MAY come from any source available to the client application.`

Does this mean that the server can provide default values for variables in the uri-template by including child data elements with matching name properties? Something like the following:

    <data 
        name="create"
        rel="http://example.org/rels/create"
        url="http://example.org/people/"
        model="g={givenName}&f={familyName}&e={email}&a={avatarUrl}"
        action="append">
        <data name="avatarUrl">
            http://example.org/defaultAvatar.png
        </data>
    </data>

Is this also possible for templated URLs?

    <data 
        name="search"
        rel="http://example.org/rels/search"
        url="http://example.org/search/(?query,pageSize)"
        action="read">
        <data name="pageSize">10</data>
    </data>

**Comment by mamund** [comment]
@thomaseizinger 

yep, the _implication_ is that a client can "find" and "associate" properties and values within the response with templated variables. there is nothing to say how this is done.

your suggestion of "child" elements seems to make the most sense. but there are other options. imagine a data element w/ a model as a stand-alone element to be used as a guide for editing individual "records" in a list of nested data elements. or a separate response that returns *only* the data-model element to be applied to other stand-alone resources that contain data elements with names/ids that match the model.

i'm open to suggestions here. this is an important thing to work out and am all ears on example implementations.

**Comment by thomaseizinger** [comment]
@mamund 

I am currently working on a Java client-library and in my experience, child elements would be relatively easy to implement. Although one could argue that this duplicates information, (The same action is possible for every item of a collection and every collection item describes the action.) I would prefer if the specification still describes several concrete ways here in order to allow compatibility between implementations. Maybe these ways could be assigned some maturity level to ease the documentation of the features a client-library supports.

Personally, I am a fan of the child elements for several reasons:
* Ease of implementation
* Readability
* Flexibility
* Similarity to HTML-forms

The only downside I can currently think of is the duplication of information which results in an increased of the response-size, however this could be overcome with the use of compression on transport-level. 

Additionally, if the definition of default values is not only possible for the `model` property but also for the `url` property, this should be stated in the specification.

**Comment by mamund** [comment]
Thomas:

NOTE: see a similar thread on the mailing list.

I see the advantage of using a "child element" approach to make
automating/parsing for form support easy and would like to see this in
action to learn how hard/easy/flexible this solution can be. This is a
CONVENTION that needs to be set up a priori between client and server
(maybe signaled via a profile string) and is definitely worth pursuing.

As i mentioned in another thread online, this is an area that was purposely
left vague in the draft spec in order to allow us to explore use-cases,
options on a wide scale w/o forcing an early STANDARD.

I'd like to see your option built-out and learn more about it. I'd also
like to see other options explored.

The sooner that happens, the sooner we'll learn of the broad range of
use-cases and possible options and will be able to converge on a viable
STANDARD.

keep me posted on your progress.

cheers.

**Comment by thomaseizinger** [comment]
FYI:
I have published my current work on the client library here:
https://github.com/rauber-projects/rauber-client

The default-value parsing for the model-string is implemented here:
https://github.com/rauber-projects/rauber-client/blob/master/src/main/java/org/rauber/action/content/UrlEncodedContentCreator.java

You can see it in action in the following unit-test:
https://github.com/rauber-projects/rauber-client/blob/master/src/test/java/org/rauber/action/content/UrlEncodedContentCreatorTest.java

I am looking forward to your feedback on the implementation :)

**Comment by mamund** [comment]
yes -- that helps.

thanks.

**Comment by thomaseizinger** [comment]
> I see the advantage of using a "child element" approach to make
automating/parsing for form support easy and would like to see this in
action to learn how hard/easy/flexible this solution can be. This is a
CONVENTION that needs to be set up a priori between client and server
(maybe signaled via a profile string) and is definitely worth pursuing.

I completely understand your point. I never intended to propose child-elements as the only reasonable way this can be done, sorry if that came across differently!

**Comment by mamund** [comment]
no-no, we're on the same page here... <g>

what i think is needed is a way to move the child-element pattern _past_
CONVENTION into STANDARD -- even if it is not the *only* STANDARD.
communication by some inline element in the representation, by a header in
the response, or maybe just a PROFILE string, etc. might be all that is
needed to properly handle this pattern.

i may be missing lots of other ideas, too.

1) we need to prove out a parsing/processing model that can recognize and
support this pattern
2) we need to come up w/ a way to safely communicate this pattern w/o
stepping on the possibility of other patterns in the future

love to see this move forward.

cheers.
